Cds 67816 pipeline rollback (#1867)
harness/developer-hub: docs/platform/8_Pipelines/define-a-failure-strategy-for-pipelines.md

---

> title: Define a failure strategy on pipelines

Reviewer: We're not supporting failure strategy on pipeline level as of now so this doc can be removed

> title: Define a failure strategy on pipelines

Author: ```suggestion
title: Rollback pipelines
```

> description: Define a failure strategy for pipelines to handle different failure conditions. 

Author: ```suggestion
description: Rollback pipelines to handle different failure conditions. 
```

---
title: Rollback pipelines
description: Rollback pipelines to handle different failure conditions. 
sidebar_position: 12
---

:::info

Currently, this feature is behind the feature flag, `PIPELINE_ROLLBACK`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.

:::

> You can add a Rollback Pipeline failure strategy to your pipelines. Rollback pipeline applies to all stages in the pipeline. 

Author: ```suggestion
You can add a Rollback Pipeline failure strategy to your pipelines. Rollback pipeline applies to all steps and stages in a pipeline. 
```

:::info

> During [Pipeline chaining](/docs/platform/pipelines/pipeline-chaining/), if there is a stage of type, pipeline, the deployment stages inside the child pipeline cannot be rolled back using the Rollback Pipeline failure strategy because the execution of the child pipeline is considered as a seperate execution.

Reviewer 2: "separate"

> During [Pipeline chaining](/docs/platform/pipelines/pipeline-chaining/), if there is a stage of type, pipeline, the deployment stages inside the child pipeline cannot be rolled back using the Rollback Pipeline failure strategy because the execution of the child pipeline is considered as a seperate execution.

Reviewer 2: If a pipeline includes a child pipeline as a stage (a.k.a. [Pipeline chaining](/docs/platform/pipelines/pipeline-chaining/)), then rolling back the parent pipeline will not roll back the child pipeline stages, since the latter is considered a separate execution. 
@namanvermah  - what will happen in this case? Only the parent will roll back? Let's specify.

Reviewer: The Pipeline Stage will be skipped in the rollback section, while the Deployment Stages in the parent Pipeline will rollback

> During [Pipeline chaining](/docs/platform/pipelines/pipeline-chaining/), if there is a stage of type, pipeline, the deployment stages inside the child pipeline cannot be rolled back using the Rollback Pipeline failure strategy because the execution of the child pipeline is considered as a seperate execution.

Author: ```suggestion
During [Pipeline chaining](/docs/platform/pipelines/pipeline-chaining/), if there is a stage of type, pipeline, the deployment stages inside the child pipeline cannot be rolled back using the Rollback Pipeline failure strategy because the execution of the child pipeline is considered as a separate execution.
```

:::

> The Rollback Pipeline failure strategy applies to the following deployments only:

Reviewer: Can we say that this is "As of now", just to indicate the support for other swimlanes is coming soon

Author: I'll say currently.

> The Rollback Pipeline failure strategy applies to the following deployments only:

Author: ```suggestion
Currently, the Rollback Pipeline failure strategy applies to the following deployments only:
```

* Kubernetes
* Native Helm
* Amazon Elastic Container Service (ECS)

> 

Reviewer 2: A subheading here? "How to set up Pipeline Rollback"?

> 1. In a pipeline, select **Advanced**.

Reviewer: Like I said earlier, we are not supporting pipeline level failure strategy, so this flow of going to Advanced section in a Pipeline is not there yet.

> 1. In a pipeline, select **Advanced**.

Author: ```suggestion
1. In a step or stage in a pipeline, select **Advanced**.
```

2. In **Failure Strategy** > **On failure of type**, select a failure type or select **All Errors**. 
   Go to [Step and Stage Failure Strategy Settings](w_pipeline-steps-reference/step-failure-strategy-settings.md) for information about error types.
3. In **Perform Action**, select **Rollback Pipeline**.

>    Here's a pipeline rollback sample YAML: 

Reviewer 2: "The equivalent YAML for this failure strategy would be:"

   ```
   failureStrategies:
     - onFailure:
         errors:
            - AllErrors                           
         action:
            type: PipelineRollback
   ```

During pipeline execution, in **Permissible Actions**, if you select **Mark as Fail**, the pipeline stage fails and rolls back. 

![](static/pipeline-failure-strategy-1.png)

> If you select **Mark as Success**, the stage passes and the next stage's execution begins. 

Author: ```suggestion
If you select **Mark as Success**, the stage passes, and the next stage's execution begins. 
```

---

:::info

Currently, this feature is behind the feature flag, `PIPELINE_ROLLBACK`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.

:::

You can add a Rollback Pipeline failure strategy to your pipelines. Rollback pipeline applies to all steps and stages in a pipeline. 

:::info

During [Pipeline chaining](/docs/platform/pipelines/pipeline-chaining/), if there is a stage of type, pipeline, the deployment stages inside the child pipeline cannot be rolled back using the Rollback Pipeline failure strategy because the execution of the child pipeline is considered as a separate execution.

:::

Currently, the Rollback Pipeline failure strategy applies to the following deployments only:
* Kubernetes
* Native Helm
* Amazon Elastic Container Service (ECS)

## Rollback pipeline steps or stages

1. In a pipeline's step or stage, select **Advanced**.
2. In **Failure Strategy** > **On failure of type**, select a failure type or select **All Errors**. 
   Go to [Step and Stage Failure Strategy Settings](w_pipeline-steps-reference/step-failure-strategy-settings.md) for information about error types.
3. In **Perform Action**, select **Rollback Pipeline**.

>    Here's a pipeline rollback sample YAML: 

Author: ```suggestion
   Here's the equivalent YAML for the Pipeline Rollback failure strategy: 
```

---
title: Rollback pipelines
description: Rollback pipelines to handle different failure conditions. 
sidebar_position: 12
---

:::info

Currently, this feature is behind the feature flag, `PIPELINE_ROLLBACK`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.

:::

You can add a Rollback Pipeline failure strategy to your pipelines. Rollback pipeline applies to all steps and stages in a pipeline. 

:::info

> If a pipeline includes a child pipeline as a stage ([Pipeline chaining](/docs/platform/pipelines/pipeline-chaining/)), rolling back the parent pipeline will not roll back the child pipeline as the latter is considered as a separate execution. In such cases, only the deployment stages of the parent pipeline rolls back.

Reviewer 2: deployment stages of the parent pipeline _roll_ back.

> If a pipeline includes a child pipeline as a stage ([Pipeline chaining](/docs/platform/pipelines/pipeline-chaining/)), rolling back the parent pipeline will not roll back the child pipeline as the latter is considered as a separate execution. In such cases, only the deployment stages of the parent pipeline rolls back.

Author: ```suggestion
If a pipeline includes a child pipeline as a stage ([Pipeline chaining](/docs/platform/pipelines/pipeline-chaining/)), rolling back the parent pipeline will not roll back the child pipeline as the latter is considered as a separate execution. In such cases, only the deployment stages of the parent pipeline roll back.
```
